### PR TITLE
Clients(Authorization): Evaluate tab updates

### PR DESCRIPTION
**Pull request by jenny-s51**
## Motivation

Extends work done on #1861 - these fixes are necessary in order to implement results screen https://marvelapp.com/prototype/5eg9ic7/screen/83188838

## Brief Description

Fixes `evaluate` call which now POSTs successfully, adds 'revert' functionality to form, fixes bug in realm role attributes and  resolves some issues with data types in `AttributeInput`

## Verification Steps

1. Create a OIDC client with Client Authentication and Authorization switches enabled.
2. Go to the new client -> Authorization -> Evaluate
3. Fill in form data and verify that the `evaluate` call POSTs successfully

## Checklist:

- [x] Code has been tested locally by PR requester
- [x] User-visible strings are using the react-i18next framework (useTranslation)
- [x] Help has been implemented
- [x] axe report has been run and resulting a11y issues have been resolved
- [x] Unit tests have been created/updated

### CODE REVIEW

File: src/clients/authorization/AuthorizationEvaluate.tsx [hunk 1]
```diff
@@ -296,11 +328,23 @@ export const AuthorizationEvaluate = ({
               helperTextInvalid={t("common:required")}
               fieldId={name!}
             >
-              <AttributeInput
-                selectableValues={resources.map((item) => item.name!)}
-                resources={resources}
-                isKeySelectable
-                name="resources"
+              <Controller
```
Comment (edewit) [inline src/clients/authorization/AuthorizationEvaluate.tsx:331]: Why would you need a `<Controller` around this control, isn't it already part of the "form"

File: src/clients/authorization/AuthorizationEvaluate.tsx [hunk 2]
```diff
@@ -28,6 +28,16 @@ import { defaultContextAttributes } from "../utils";
...
 import type ScopeRepresentation from "@keycloak/keycloak-admin-client/lib/defs/scopeRepresentation";
+import type { KeyValueType } from "../../components/attribute-form/attribute-convert";
+
+interface EvaluateFormInputs
+  extends Omit<ResourceEvaluation, "context" | "resources"> {
+  applyToResource: boolean;
+  alias: string;
+  authScopes: string[];
+  context: { [key: string]: string };
+  resources: { [key: string]: string }[];
```
Comment (jonkoops) [inline src/clients/authorization/AuthorizationEvaluate.tsx:39]: You can use a [`Record`](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type) type here.
```suggestion
  context: Record<string, string>;
  resources: Record<string, string>[];
```

File: src/clients/authorization/AuthorizationEvaluate.tsx [hunk 3]
```diff
@@ -90,19 +98,37 @@ export const AuthorizationEvaluate = ({
...
-      context: formValues.context,
-      resources: formValues.resources,
-      clientId: selectedClient?.id!,
-    };
-    return adminClient.clients.evaluateResource(
-      { id: clientId!, realm: realm.realm },
-      resEval
-    );
+  const evaluate = async () => {
+    if (await trigger()) {
```
Comment (jonkoops) [inline src/clients/authorization/AuthorizationEvaluate.tsx:102]: You can reduce nesting here by using an [early return pattern](https://medium.com/swlh/return-early-pattern-3d18a41bba8).

```ts
if (!(await trigger())) {
  return;
}

// The rest of the code.
```

File: src/clients/authorization/AuthorizationEvaluate.tsx [hunk 4]
```diff
@@ -90,19 +98,37 @@ export const AuthorizationEvaluate = ({
...
+      const formValues = form.getValues();
+      const resEval: ResourceEvaluation = {
+        roleIds: formValues.roleIds ?? [],
+        clientId: selectedClient ? selectedClient.id! : clientId,
+        userId: selectedUser?.id!,
+        resources: resources.filter((resource) =>
+          form
+            .getValues()
+            .resources.map((item) => item.key)
```
Comment (jonkoops) [inline src/clients/authorization/AuthorizationEvaluate.tsx:111]: This logic is repeated for every resource you are iterating, it might be a good idea to store it in a variable. You also should not have to call `getValues()` again as there is already a variable that contains this called `formValues` above.

```ts
const keys = formValues.resources.map(({ key }) => key);
```

File: src/clients/authorization/AuthorizationEvaluate.tsx [hunk 5]
```diff
@@ -90,19 +98,37 @@ export const AuthorizationEvaluate = ({
...
+        entitlements: false,
+        context: {
+          attributes: formValues.context.length
+            ? {
+                [Object.entries(formValues.context).map(
+                  (item) => (item[1] as unknown as KeyValueType).key
+                ) as unknown as string]: Object.entries(formValues.context).map(
+                  (item) => (item[1] as unknown as KeyValueType).value
+                )[0],
+              }
```
Comment (jonkoops) [inline src/clients/authorization/AuthorizationEvaluate.tsx:123]: This code is very hard to parse, but looking at the type casting this seems to entirely change the type of `formValues.context`. Why does this need to happen?

I'd like to see this code without type-casting, as this disables certain compiler checks that can prevent errors during runtime.

File: src/clients/authorization/AuthorizationEvaluate.tsx [hunk 6]
```diff
@@ -296,11 +328,23 @@ export const AuthorizationEvaluate = ({
...
+              <Controller
+                name="applyToResource"
+                defaultValue=""
+                control={control}
+                render={() => (
+                  <AttributeInput
+                    selectableValues={
+                      resources
+                        .map((item) => [{ name: item.name }])
+                        .flat() as AttributeType[]
```
Comment (jonkoops) [inline src/clients/authorization/AuthorizationEvaluate.tsx:340]: There is no need to call `flat()` here or to type-cast the items if you map it directly to an object instead of an array with a generic type.

```ts
resources.map<AttributeType>((item) => ({
  name: item.name!,
}))
```

According to the `AttributeType` type you will also have to provide a `key` which is missing here. Because this was being type-casted this error was suppressed by the compiler.

File: src/clients/authorization/AuthorizationEvaluate.tsx [hunk 7]
```diff
@@ -390,35 +434,34 @@ export const AuthorizationEvaluate = ({
...
+            data-testid="authorization-revert"
+            variant="link"
+            onClick={() => reset()}
+          >
+            {t("common:revert")}
+          </Button>
+          <Button
+            data-testid="authorization-revert"
+            variant="primary"
+            onClick={() => reset}
```
Comment (jonkoops) [inline src/clients/authorization/AuthorizationEvaluate.tsx:459]: You probably want to call the `reset()` function here.
```suggestion
            onClick={() => reset()}
```

File: src/components/attribute-input/AttributeInput.tsx [hunk 8]
```diff
@@ -56,7 +58,7 @@ export const AttributeInput = ({
     if (!fields.length) {
       append({ key: "", value: "" });
     }
-  }, []);
+  }, [reset]);
```
Comment (jonkoops) [inline src/components/attribute-input/AttributeInput.tsx:61]: Why does this effect now depend on `reset`? It doesn't look like this dependency is used inside of the effect at all.

**Review by jonkoops — CHANGES_REQUESTED** [review]

File: src/clients/authorization/AuthorizationEvaluate.tsx (hunk 2 above)
Comment (jenny-s51) [reply src/clients/authorization/AuthorizationEvaluate.tsx:39]: Good call - thanks @jonkoops

File: src/clients/authorization/AuthorizationEvaluate.tsx (hunk 5 above)
Comment (jenny-s51) [reply src/clients/authorization/AuthorizationEvaluate.tsx]: Nice catch, I've updated this logic - the submitted types should be correct and casting is no longer necessary

File: src/clients/authorization/AuthorizationEvaluate.tsx (hunk 1 above)
Comment (jenny-s51) [reply src/clients/authorization/AuthorizationEvaluate.tsx:329]: Thanks! Removed 👍

File: src/components/attribute-input/AttributeInput.tsx (hunk 8 above)
Comment (jenny-s51) [reply src/components/attribute-input/AttributeInput.tsx:63]: The reset sets `fields.length` to 0 -- if it is not a dependency, the field is removed rather than being set to empty

File: src/clients/authorization/AuthorizationEvaluate.tsx (hunk 5 above)
Comment (jonkoops) [reply src/clients/authorization/AuthorizationEvaluate.tsx]: Awesome, nice work!

File: src/clients/authorization/AuthorizationEvaluate.tsx [hunk 9]
```diff
@@ -90,15 +110,31 @@ export const AuthorizationEvaluate = ({
     []
   );
 
-  const evaluate = (formValues: ResourceEvaluation) => {
+  console.log(form.getValues());
```
Comment (jonkoops) [inline src/clients/authorization/AuthorizationEvaluate.tsx:113]: Stray console log statement here.
```suggestion
```

File: src/clients/authorization/AuthorizationEvaluate.tsx [hunk 10]
```diff
@@ -90,15 +110,31 @@ export const AuthorizationEvaluate = ({
...
+      resources: resources.filter((resource) => keys.includes(resource.name!)),
       entitlements: false,
-      context: formValues.context,
-      resources: formValues.resources,
-      clientId: selectedClient?.id!,
+      context: {
+        attributes: Object.fromEntries(
+          formValues.context.attributes
+            .filter((item) => item.key || item.value !== "")
+            .map(Object.values)
```
Comment (jonkoops) [inline src/clients/authorization/AuthorizationEvaluate.tsx:130]: Not sure if this works correctly, if I split this code out then `atrributes` gets an `any` type. You might want to make this a bit more explicit.
```suggestion
            .map(({ key, value }) => [key, value])
```

File: src/components/attribute-input/AttributeInput.tsx [hunk 11]
```diff
@@ -52,11 +54,13 @@ export const AttributeInput = ({
     name,
   });
 
+  console.log("huh", fields.length);
```
Comment (jonkoops) [inline src/components/attribute-input/AttributeInput.tsx:57]: Another stray console statement here.
```suggestion
```

**Review by jonkoops — CHANGES_REQUESTED** [review]

**Review by jonkoops — APPROVED** [review]

**Comment by jenny-s51** [comment]
All Cypress tests passing locally ✅
